File: gc-auth/gc-auth-security/src/main/java/com/gc/auth/security/controller/LoginController.java
package com.gc.auth.security.controller;

import com.gc.auth.security.pojo.dto.UserLoginDTO;
import com.gc.auth.security.service.AuthService;
import com.gc.common.auth.constants.LoginTypeConstants;
import com.gc.common.auth.core.RestUserDetails;
import com.gc.common.auth.model.LoginResult;
import com.gc.common.auth.model.RestUserDetailsImpl;
import com.gc.common.base.message.Result;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.ResponseBody;

import javax.servlet.http.HttpServletRequest;
import javax.validation.Valid;

/**
 * jwt登陆
 * @author jackson
 * 2020/2/14 10:22 下午
 */
@RequestMapping
@ResponseBody
public class LoginController {


    private AuthenticationManager authenticationManager;

    private AuthService authService;


    public LoginController(AuthenticationManager authenticationManager, AuthService authService) {
        this.authenticationManager = authenticationManager;
        this.authService = authService;
    }

    /**
     * 登陆接口
     * @param parameter 登录参数
     * @return
     */
    @PostMapping("public/auth/login")
    public Result<LoginResult> login(@RequestBody @Valid UserLoginDTO parameter) {
        return Result.success(this.doLogin(parameter, LoginTypeConstants.WEB));
    }

    /**
     * 移动端登录接口
     * @param parameter 登录参数
     * @return
     */
    @PostMapping("public/auth/mobileLogin")
    public Result<LoginResult> mobileLogin(@RequestBody @Valid UserLoginDTO parameter) {
        return Result.success(this.doLogin(parameter, LoginTypeConstants.MOBILE));
    }

    /**
     * 登出接口
     * @return
     */
    @PostMapping("auth/logout")
    public Result<Object> logout(HttpServletRequest request) {
        this.authService.logout(request);
        return Result.success(null, "登出成功");
    }


    /**
     * 执行登陆
     * @param parameter 登录参数
     * @return
     */
    private LoginResult doLogin(UserLoginDTO parameter, LoginTypeConstants type) {
        final Authentication authentication = this.authenticationManager.authenticate(new UsernamePasswordAuthenticationToken(parameter.getUsername(), parameter.getPassword()));

        final RestUserDetails userDetails = (RestUserDetailsImpl) authentication.getPrincipal();
        SecurityContextHolder.getContext()
                .setAuthentication(authentication);
        final String jwt = this.authService.doLogin(authentication, parameter.getRemember(), type);
        return LoginResult.builder()
                .user(userDetails)
                .token(jwt)
                .roles(userDetails.getRoles())
                .permissions(userDetails.getPermissions())
                .build();
    }

}

File: smart-module-starter/smart-auth/src/main/kotlin/com/smart/auth/controller/LoginController.kt
package com.smart.auth.controller

import com.smart.auth.AuthProperties
import com.smart.auth.common.constants.AuthConstants
import com.smart.auth.common.utils.AuthUtils
import com.smart.auth.model.po.LoginPO
import com.smart.common.message.Result
import com.smart.common.utils.IPUtils
import com.smart.system.service.SysUserService
import org.apache.shiro.SecurityUtils
import org.apache.shiro.authc.UsernamePasswordToken
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpHeaders
import org.springframework.util.StringUtils
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import javax.servlet.http.HttpServletRequest

/**
 * 登录接口
 * @author ming
 * 2019/6/21 下午3:00
 */
@RestController
@RequestMapping
class LoginController {

    companion object {
        private val LOGGER = LoggerFactory.getLogger(LoginController :: class.java)
    }

    @Autowired
    private lateinit var authProperties: AuthProperties

    @Autowired
    private lateinit var userService: SysUserService


    /**
     * 登录接口
     */
    @PostMapping("/public/login")
    fun login(@RequestBody user: LoginPO, request: HttpServletRequest): Result<Map<String, Any>?> {
        return this.doLogin(request, user.username, user.password, AuthConstants.LOGIN_WEB_TYPE)
    }

    /**
     * 移动端登录接口
     */
    @PostMapping("/public/mobileLogin")
    fun mobileLogin(@RequestBody user: LoginPO, request: HttpServletRequest): Result<Map<String, Any>?> {
        return this.doLogin(request, user.username, user.password, AuthConstants.LOGIN_MOBILE_TYPE)
    }

    /**
     * 执行登录
     * @param 登录类型（web、移动）
     */
    private fun doLogin(request: HttpServletRequest, username: String?, password: String?, type: AuthConstants): Result<Map<String, Any>?> {
        if (StringUtils.isEmpty(username) || StringUtils.isEmpty(password)) {
            LOGGER.warn("用户名或密码为空，username：{}  password：{}", username, password)
            return Result.failure(403, "用户名或密码不能为空")
        }
        return try {
            // 执行登录
            SecurityUtils.getSubject().login(UsernamePasswordToken(username, password))
            val session = SecurityUtils.getSubject().session
            // 设置session的超时事件
            if (type == AuthConstants.LOGIN_WEB_TYPE) {
                session.timeout = this.authProperties.session.timeout.global * 1000
            } else if (type == AuthConstants.LOGIN_MOBILE_TYPE) {
                session.timeout = this.authProperties.session.timeout.mobile * 1000
            }
            val permissionList = this.queryUserPremissionList()
            session.setAttribute(AuthConstants.PERMISSION, permissionList)
            session.setAttribute(AuthConstants.LOGIN_IP, IPUtils.getIpAddr(request))
            // 返回数据
            val user = AuthUtils.getCurrentUser()
            user?.let {
                it.password = null
            }
            val data = mutableMapOf(
                    HttpHeaders.AUTHORIZATION to session.id.toString(),
                    "permission" to permissionList
            )
            user?.let {
                it.password = null
                data["user"] = it
            }
            Result.success(data)
        } catch (e: Exception) {
            e.printStackTrace()
            Result.failure(e.message)
        }
    }

    /**
     * 查询用户权限信息
     */
    private fun queryUserPremissionList(): Set<String> {
        val userId = AuthUtils.getCurrentUserId() ?: return setOf()
        val premissionMap: Map<String, Set<String>> = this.userService.queryPermissionList(listOf(userId))
        return premissionMap[userId] ?: setOf()
    }
}